File: 206 Reverse Linked List/reverseList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        # "None" is enough. If use ListNode(None), the result will be [None] instead of [] for input []
        pre = None
        cur = head
        
        while cur != None :
            next_one = cur.next
            
            cur.next = pre
            pre = cur
            cur = next_one
            
        return pre
        

File: 剑指offer/反转链表/ReverseList.cpp
// 运行时间：4ms 占用内存：612k
class Solution {
public:
    ListNode* ReverseList(ListNode* pHead) {
        ListNode* pre = NULL, *cur = pHead, *next = NULL;
        while( cur ){
            next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        }
        return pre;
    }
};

/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/